src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs

```diff
@@ -199,58 +173,87 @@ private static bool AnyTopLevelTypeMatchesDocumentName(State state, Cancellation
 
-            // qualified type name `Outer.Inner` matches file names `Inner.cs` and `Outer.Inner.cs`
-            return typeNameParts.SequenceEqual(fileNameParts, StringComparer.CurrentCulture);
+        if (typeNameParts.Length != fileNameParts.Length)
+            return false;
+
+        // qualified type name `Outer.Inner` matches file names `Inner.cs` and `Outer.Inner.cs` as well as
+        // Outer`1.Inner`2.cs
+        for (int i = 0, n = typeNameParts.Length; i < n; i++)
+        {
+            if (!TypeNameMatches(fileNameParts[i], typeNameParts[i].name, typeNameParts[i].arity))
+                return false;
         }
 
-        return namesMatch;
+        return true;
+    }
+
+    private static bool TypeNameMatches(string documentNameWithoutExtension, string typeName, int arity)
+    {
+        if (typeName.Equals(documentNameWithoutExtension, StringComparison.CurrentCulture))
+            return true;
+
+        if ($"{typeName}`{arity}".Equals(documentNameWithoutExtension, StringComparison.CurrentCulture))
+            return true;
+
+        return false;
     }
 
-    private static ImmutableArray<string> GetSuggestedFileNames(
+    private ImmutableArray<string> GetSuggestedFileNames(
         TTypeDeclarationSyntax typeNode,
-        bool isNestedType,
-        string typeName,
         string documentNameWithExtension,
-        SemanticModel semanticModel,
-        CancellationToken cancellationToken)
+        bool includeArity)
     {
+        var isNestedType = IsNestedType(typeNode);
+        var (typeName, arity) = this.GetSymbolNameAndArity(typeNode);
         var fileExtension = Path.GetExtension(documentNameWithExtension);
 
         var standaloneName = typeName + fileExtension;
 
-        // If it is a nested type, we should match type hierarchy's name parts with the file name.
+        using var _ = ArrayBuilder<string>.GetInstance(out var suggestedFileNames);
+
+        suggestedFileNames.Add(typeName + fileExtension);
+        if (includeArity && arity > 0)
+            suggestedFileNames.Add($"{typeName}`{arity}{fileExtension}");
+
         if (isNestedType)
         {
-            var typeNameParts = GetTypeNamePartsForNestedTypeNode(typeNode, semanticModel, cancellationToken);
-            var dottedName = typeNameParts.Join(".") + fileExtension;
+            var typeNameParts = GetTypeNamePartsForNestedTypeNode(typeNode);
+            AddNameParts(typeNameParts.Select(t => t.name));
 
-            return [standaloneName, dottedName];
+            if (includeArity && typeNameParts.Any(t => t.arity > 0))
+                AddNameParts(typeNameParts.Select(t => t.arity > 0 ? $"{t.name}`{t.arity}" : t.name));
         }
-        else
+
+        return suggestedFileNames.ToImmutableAndClear();
+
+        void AddNameParts(IEnumerable<string> parts)
+        {
+            AddNamePartsWithSeparator(parts, ".");
+            AddNamePartsWithSeparator(parts, "+");
```

Comment:
moved to also detect the pattern of `X+Y.cs` to represent a nested type.  We use that in a few places in roslyn's codebase itself.